### PR TITLE
chore(swingset): improve typechecks

### PR DESCRIPTION
**Pull request by warner**
This enables typechecking for a few more swingset modules.

### CODE REVIEW

File: packages/SwingSet/src/kernel/vatTranslator.js
Comment (FUDCo) [inline packages/SwingSet/src/kernel/vatTranslator.js:158]: I like the "before" code that destructures into named consts much better than the "after" code that directly accesses `kd` as an array.  It does a superior job of communicating intent.  I advocate reverting this change.

File: packages/SwingSet/src/kernel/vatTranslator.js
Comment (FUDCo) [inline packages/SwingSet/src/kernel/vatTranslator.js:482]: See prior comment about uglification in `kernelDeliveryToVatDelivery`.  Same thing here.

**Review by FUDCo — CHANGES_REQUESTED** [review]
There's a couple of bits of legibility disimprovement that I'd really like to see reverted, but other than that it looks good once TypeScript is happy with the state of things.

File: packages/SwingSet/src/kernel/vatTranslator.js
Comment (warner) [reply packages/SwingSet/src/kernel/vatTranslator.js:158]: I hate it too, but it was the only way @michaelfig and I could find to appease the typechecker. The core issue (if I understand properly) is that typescript isn't clever enough to combine `const [a,b] = c;` and a subsequent constraint on `a` into a deduced constraint on `c`. So even though the `case 'message'` tells us that `type` must be `message`, and therefore `kd` must be a `KernelDeliveryMessage`, the typechecker doesn't realize that. But `kd[0] === 'message'` *is* enough to provoke that deduction.

The same sort of issue is what requires the second part: once it knows that `kd` is a `KernelDeliveryMessage` (within this one `case`), it knows what `kd[1]` is, and `const [_, ...args] = kd` allows it to know what `args` is. But if the factored out `[type, args] = kd` is before the switch, it *can't* deduce a type for `args`, and then all the calls to the more specific functions look like they're using the wrong (or an unknown) type.

This is clearly a tension between human readability and typechecker readability. I guess the deciding question is whether appeasing the typechecker gives us enough advantage elsewhere to make up for the impact to human readability. In this case, the code is ugly, but not unreadably ugly, so I think it's worth it.

But I'm really hoping that the typechecker gets smarter some day, so we can go back to the old code, because I agree, this looks gross.

Comment (FUDCo) [reply packages/SwingSet/src/kernel/vatTranslator.js:158]: This is something that TypeScript can do in "native" TypeScript code, so I'm not convinced we can't make it work here.  I'm inclined to whine louder about this.  If TypeScript is forcing us to write obviously worse code just to satisfy the type checker then something is very wrong.  If we are going to be pushed into something that's harder to read, I would say that if a type annotation comment above the `const [a, b] = c; ` line could do the trick than I could live with that.  But this is just horrible.

**Comment by warner** [comment]
ugh, I have to battle with the typechecker again

File: packages/SwingSet/src/kernel/vatTranslator.js
Comment (turadg) [reply packages/SwingSet/src/kernel/vatTranslator.js:154]: It makes sense that the earlier case didn't work because it couldn't know the length of `args` before the switch. Putting it under the case lets it constrain that.

Though it would be more more readable to ditch the _rest_,
```
 function kernelDeliveryToVatDelivery(kd) {
    switch (kd[0]) {
      case 'message': {
        return translateMessage(kd[1], kd[2]);
```


[playground example](https://www.typescriptlang.org/play?jsx=0&ts=4.2.3&ssl=29&ssc=1&pln=29&pc=2#code/C4TwDgpgBAwghmAInYcA8BlANge2APigF4oBvAWACgoaoAjHAExAC4oBnYAJwEsA7AOYAaKrQ65g7NtjwBtALojKAXwDcVUJCgYA7vwEYIweEhRxisBMlRpOvQfg3hoAWQjt2cAdBIVqAWyMACyY2O30lKDguASltPUFDYyszSK53AFcsYAB+MO4IqjUnLQBpCC4+CCxEap4ANwqQNw8vHyhZVAE2AHJA1u8eoShUGKN8+2Eof3ZuqBbPb3l1Sk1ocsrq2qwGpoA5PB4AMxALTq9evkOToah09hwsDOAeHD448MEFKGWqEvWKlUanVGlwQAB5OgAKwgAGNgBYNkDtrswQs2lAAD5QJFbEH7a4gFZURhwrDRaBHDJ8eGvPgjLhwd7k4AQdHeAAUo28wAmEWmszY7IgAEo2EyiSSyRSoFSaS83gymewWRADi8Thz7o9nnSPgUvvIxVE+JLKFRYW9OFAANaAvE7UEgAAqOAAaigUU6LBybYw2LjgY6mpCYfCRcRCH5RLR2HpgLCglBfYxZAAGI1kGNiWFwdjQPruRYQHosbNiO5GDKVJXMlBsottFOyACMiltqYATEaVhWoLn81AelcNSBS+WxOlgNX6dxlar1ccQM22yLexXSUc4FleRPaMAglwcDooFUTwBRLhHrgcgAG1JtVx09LtmyDqJAADo1lAACSkP103kZRbzXctlCKIA) (building off @mhofman's)

Comment (turadg) [reply packages/SwingSet/src/kernel/vatTranslator.js:158]: > This is something that TypeScript can do in "native" TypeScript code

You mean in `.ts` files? What does it look like there?

Comment (FUDCo) [reply packages/SwingSet/src/kernel/vatTranslator.js:158]: I don't recall, as I wasn't the one who did the TypeScript voodoo.  We had a guy who was a savant w.r.t H-M type systems.  I just remember we had a bunch of this kind of stuff.

**Review by FUDCo — APPROVED** [review]
At @warner's request I'm clearing my request for changes and have resigned myself to the TS-driven uglification.

File: packages/SwingSet/src/vats/comms/state.js
Comment (turadg) [inline packages/SwingSet/src/vats/comms/state.js]: Looks like `!s` is to handle the null case, but it also handles the empty string case.
```suggestion
  if (!s) {
```

**Review by turadg — APPROVED** [review]

File: packages/SwingSet/src/vats/comms/state.js
Comment (warner) [reply packages/SwingSet/src/vats/comms/state.js]: Oops, thanks, I'm perpetually confused about JS's truthyness rules. In this case I think I'll add a comment for others like me. The function might get `undefined` (because the DB key that feeds it was not present, and `undefined` is the obvious return value from a `.get(missingKey)`, or `null` (because the DB lookup *actually* happened on the other side of a JSON-encoded vat-worker-to-kernel-process pipe and the JSON decoder produces `null` instead of `undefined`), or `''` (because the key is present but was populated with `slots.join(',')` and `slots` was empty). In all three cases, and empty list is the correct return value.

**Comment by warner** [comment]
I'm going to wait for #4420 to land before landing this one

**Comment by warner** [comment]
I think this should not conflict with #4442

**Comment by mhofman** [comment]
> I think this should not conflict with #4442

That was subsumed by #4420

**Pull request merged**
